firestore/bulkwriter_test.go

```diff
@@ -0,0 +1,233 @@
+						Delete: docPrefix + "c",
+					},
+				},
+			},
+		},
+		&pb.BatchWriteResponse{
+			WriteResults: []*pb.WriteResult{
+				{UpdateTime: aTimestamp3},
+			},
+			Status: []*status.Status{
+				{
+					Code:    int32(codes.OK),
+					Message: "delete test successful",
+				},
+			},
+		},
+	)
+
+	// Update
+	srv.addRPC(
+		&pb.BatchWriteRequest{
+			Database: c.path(),
+			Writes: []*pb.Write{
+				{
+					Operation: &pb.Write_Update{
+						Update: &pb.Document{
+							Name:   docPrefix + "f",
+							Fields: map[string]*pb.Value{"*": intval(3)},
+						},
+					},
+					UpdateMask: &pb.DocumentMask{FieldPaths: []string{"`*`"}},
+					CurrentDocument: &pb.Precondition{
+						ConditionType: &pb.Precondition_Exists{
+							Exists: true,
+						},
+					},
+				},
+			},
+		},
+		&pb.BatchWriteResponse{
+			WriteResults: []*pb.WriteResult{
+				{UpdateTime: aTimestamp3},
+			},
+			Status: []*status.Status{
+				{
+					Code:    int32(codes.OK),
+					Message: "update test successful",
+				},
+			},
+		},
+	)
+
+	ctx := context.Background()
+	bw := c.BulkWriter(ctx)
+	wantWRs := []*WriteResult{{aTime}, {aTime2}, {aTime3}, {aTime3}}
+	tcs := []bulkwriterTestCase{
+		{
+			name: "Create()",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				return b.Create(c.Doc("C/a"), testData)
+			},
+		},
+		{
+			name: "Set()",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) { return b.Set(c.Doc("C/b"), testData) },
+		},
+		{
+			name: "Delete()",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				return b.Delete(c.Doc("C/c"))
+			},
+		},
+		{
+			name: "Update()",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				return b.Update(c.Doc("C/f"), []Update{{FieldPath: []string{"*"}, Value: 3}})
+			},
+		},
+	}
+
+	for i, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			j, err := tc.test(bw)
+			if err != nil {
+				t.Errorf("bulkwriter: cannot call %s for document\n", tc.name)
+			}
+			if j == nil {
+				t.Errorf("bulkwriter: got nil WriteResult for call to %s\n", tc.name)
+			}
+
+			bw.Flush()
+
+			wr, err := j.Results()
+			if err != nil {
+				t.Errorf("bulkwriter:\nwanted %v,\n, got error: %v", wantWRs[i], err)
+			}
+
+			if !testEqual(wr, wantWRs[i]) {
+				t.Errorf("bulkwriter:\nwanted %v,\n got %v\n", wantWRs[i], wr)
+			}
+		})
+	}
+}
+
+func TestBulkWriterErrors(t *testing.T) {
+	c, _, cleanup := newMock(t)
+	defer cleanup()
+	ctx := context.Background()
+	b := c.BulkWriter(ctx)
+
+	tcs := []bulkwriterTestCase{
+		{
+			name: "empty document reference",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				return b.Delete(nil)
+			},
+		},
+		{
+			name: "cannot write to same document twice",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				b.Create(c.Doc("C/a"), testData)
+				return b.Delete(c.Doc("C/a"))
+			},
+		},
+		{
+			name: "cannot ask a closed BulkWriter to write",
+			test: func(b *BulkWriter) (*BulkWriterJob, error) {
+				b.cancel()
```

Comment:
Shouldn't this be called within `b.End()`?

Reply:
I don't think so. `Bundler.Flush()`, which is called from within `End()`, isn't blocking. If we attempt to call `cancel()` from within `End()`, then we have a race condition that results in the last few write attempts being canceled.

Reply:
Actually, wait a minute. Correcting myself ...

I'm not sure whether `Bundler.Flush()` blocks or not. Local tests seem to indicate that it doesn't: when I put `cancel()` within `End()` and then call `End()`, I see that the integration tests fail with "context canceled" errors for each write.